### PR TITLE
C++: Implement `clearsContent`

### PR DESCRIPTION
**Pull request by MathiasVP**
Until now it hasn't been possible to implement `clearsContent` for C/C++ since the predicate requires use-use flow. See [here](https://github.com/github/codeql/blob/1d7efd8e822c3890ad6444fb9d34fa6bec4de8b9/docs/ql-libraries/dataflow/dataflow.md#field-flow-barriers):
> However, this relies on the local step relation using the smallest possible use-use steps. If local flow is implemented using def-use steps, then clearsContent might not be easy to use.

but now that we have use-use flow we can pretty much just take the implementation from the documentation 🎉.

It required a few modifications because of our precise handling of indirections, though. For example, the following shouldn't produce an alert:
```cpp
int source();
void sink(int);
struct S { int** p; };

void test() {
  S s;
  **s.p = source();
  *s.p = nullptr;
  sink(**s.p);
}
```
but this should produce one:
```cpp
int* source();
void sink(int*);
struct S { int** p; };

void test() {
  S s;
  *s.p = source();
  **s.p = 0;
  sink(*s.p);
}
```

The tests added in e7f187c933c4751fb53e08a06061be1d4ef414d6 should hopefully demonstrate that this is the behavior we get from this implementation.

### CODE REVIEW

**Comment by MathiasVP** [comment]
I've started a DCA run for this now. I expect that it'll show a bunch of lost results since the effect of this PR is really to _reduce_ the amount of flow we're getting.

**Review by jketema — APPROVED** [review]
LGTM, but I see some result changes in DCA that need to be looked at.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:628]:
```suggestion
 * The boolean `certain` is true if the destination address does not involve
```

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll]: Thanks!

**Comment by jketema** [comment]
> LGTM, but I see some result changes in DCA that need to be looked at.

I retracted this, as you pushed some significant changes 😄

**Comment by MathiasVP** [comment]
> > LGTM, but I see some result changes in DCA that need to be looked at.
> 
> I retracted this, as you pushed some significant changes 😄

That's fair 😅. Indeed, the FN I added in https://github.com/github/codeql/pull/12344/commits/f19ffcefdba731e11acd38203a01cf5e20900a1d is a reduced example of some of the lost results I saw on a project. I've started a new DCA run now 🤞.

**Review by jketema — APPROVED** [review]
LGTM (pending DCA results)

**Comment by MathiasVP** [comment]
The semantics we need to capture here needs to be thought about more thoroughly before we merge this, I think. I've added several new reduced FNs in https://github.com/github/codeql/pull/12344/commits/b4c73d7f920f85fd59e0b2754d3dabddc3c9ad48 from the lost TPs on DCA. I'll park this PR for now, and get back to it after we've merged the feature branch.

**Pull request closed without merging**
